crawler/crawl_emap.py

```diff
@@ -0,0 +1,130 @@
+                        sleep(delay)
+                        retries += 1
+                    else:
+                        raise
+        return wrapper
+    return decorator
+
+def get_auth_token():
+    try:
+        parser = configparser.ConfigParser()
+        config_filepath = CONFIG_FILE
+        parser.read(config_filepath)
+        return parser['auth']['token']
+    except Exception as ex:
+        raise ValueError(
+            f'Failed to retrieve watttime credentials: {ex}') from ex
+
+
+def prepare_insert_query(data):
+    carbon_intensity = data.get('carbonIntensity', 0.0)
+    if carbon_intensity is None:
+        carbon_intensity = 0.0
+
+    low_carbon_percentage = 0.00
+    renewable_percentage = 0.00
+
+    datetime = data['datetime']
+
+    zone_id = data['zone']
+
+    query = """
+    INSERT INTO EMapCarbonIntensity(DateTime, ZoneId, CarbonIntensity, LowCarbonPercentage, RenewablePercentage)
+    VALUES (%s, %s, %s, %s, %s)
+    ON CONFLICT (DateTime, ZoneId) DO UPDATE SET
+    CarbonIntensity = EXCLUDED.CarbonIntensity,
+    LowCarbonPercentage = EXCLUDED.LowCarbonPercentage,
+    RenewablePercentage = EXCLUDED.RenewablePercentage;
+    """
+
+    return query, (datetime, zone_id, carbon_intensity, low_carbon_percentage, renewable_percentage)
+
+@exponential_backoff(should_retry=lambda ex: ex.response.status_code == 429)
+def fetch(zone):
+    url = f"https://api-access.electricitymaps.com/free-tier/carbon-intensity/history?zone={zone}"
+    headers = {
+        "auth-token": get_auth_token()
+    }
+
+    history_response = requests.get(url, headers=headers)
+
+    history_response_json = history_response.json()
+    try:
+        assert history_response.ok, f"Error in EMap request for getting history according to zone ({history_response.status_code}): {history_response.text}"
+        # Rest of your code goes here, which will execute if the assertion passes
+    except AssertionError as error:
+        print(str(error), file=sys.stderr)
+
+    return history_response_json
+
+
+def update(conn, history_response_json):
+    if 'history' in history_response_json:
+        for history in history_response_json['history']:
+            query, data = prepare_insert_query(history)
+            with conn, conn.cursor() as cur:
+                try:
+                    cur.execute(query, data)
+                except psycopg2.Error as ex:
+                    print(datetime.now().isoformat(),
+                          f"Failed to execute set_last_updated query.: {ex}",
+                          file=sys.stderr)
+
+
+def fetch_and_update(conn, zone):
+    history_response_json = fetch(zone)
+
+    update(conn, history_response_json)
+
+
+def get_all_electricity_zones():
+    url = "https://api-access.electricitymaps.com/free-tier/zones"
+    headers = {
+        "auth-token": get_auth_token()
+    }
+
+    zone_response = requests.get(url, headers=headers)
+
+    zone_response_json = zone_response.json()
+
+    try:
```

Comment:
Generally it's not necessary to try except and then immediately print the error itself. It's not necessary unless you're crossing function boundary. Because here it's completely equivalent to
```
if not zone_response.ok:
    print("Error in ...")
```

Reply:
Fixed, the try except has been removed.